stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp

```diff
@@ -59,6 +71,15 @@ struct SetTemperature {
     using ParseResult = std::optional<SetTemperature>;
     float temperature;
 
+    static auto write_response_into(char* buf, size_t available) -> size_t {
```

Comment:
I'm curious if there is a way (or whether it makes sense) to make `write_response_into` be more generic/shared among the structs. For any structs that require additional parameters in their response, maybe we could do a separate function which appends those parameters.

Reply:
yeah actually this is the one case where going commit-by-commit kind of messes up. if you look at the second-to-last commit, https://github.com/Opentrons/opentrons-modules/pull/134/commits/5d220d61d17934503fd20d06e00d4209d2c9e5b8 (which switches to using iterator pairs) it boils down to a single call to `write_string_to_iterpair` which I think is ok.

the other way to do it would be free function template specializations i'm pretty sure:

```c++
// declare-only for the fully-unstructured template since we don't want to ever actually use it,
// only the specializations
template<typename InputIt, typename Limit, typename GCode, typename... Args>
auto write_response_into(Inputit input, Limit limit, GCode gcode, Args... args) -> InputIt;

template<typename InputIt, typename Limit, typename GCode>
requires std::is_same_v<GCode, M123>  // Limit this specialization to one of a couple gcodes.
             || std::is_same_v<GCode, M104>  // could also be done by giving gcodes a type tag,
                                                                    // a using-declaration called like no_arguments or
                                                                    // something
auto write_response_into(Inputit input, Limit limit, GCode gcode) -> InputIt {
   // do the no-args case; take advantage of some method on the GCode object that writes the
   // gcode-specific data
}

template< InputIt, typename Limit, typename GCode>
requires std::is_same_v<GCode, ...>  // Same deal with the limit but limit it to one of the other
             || ....                                         // ones
auto write_response_into(InputIt input, Limit limit, GCode gcode, uint32_t value) -> InputIt {
   // do the int response writing, taking advantage of some other GCode message
}
```

is this better? honestly not sure.